src/bosh-director-core/lib/bosh/director/core/templates/job_instance_renderer.rb

```diff
@@ -3,28 +3,55 @@
 require 'bosh/director/formatter_helper'
 
 module Bosh::Director::Core::Templates
+
+  # @param [Array<DeploymentPlan::Job>] instance_jobs
+  # @param [JobTemplateLoader] job_template_loader
   class JobInstanceRenderer
-    def initialize(templates, job_template_loader)
-      @templates = templates
+    def initialize(instance_jobs, job_template_loader)
+      @instance_jobs = instance_jobs
       @job_template_loader = job_template_loader
     end
 
-    def render(spec)
+    # Render all templates for a Bosh instance.
+    #
+    # From a list of instance jobs (typically comming from a single instance
+    # plan, so they cover all templates of some instance) this method is
+    # responsible for orchestrating several tasks.
+    #
+    # Lazily-delegated work to a 'JobTemplateLoader' object:
+    #   - Load all templates of the release job that the instance job is
+    #     referring to
+    #   - Convert each of these to a 'JobTemplateRenderer' object
+    #
+    # Work done here on top of this:
+    #   - Render each template with the necessary bindings (comming from
+    #     deployment manifest properties) for building the special 'spec'
+    #     object that the ERB rendring code can use.
+    #
+    # The actual rendering of each template is delegated to its related
+    # 'JobTemplateRenderer' object, as created in the first place by the
+    # 'JobTemplateLoader' object.
+    #
+    # @param [Hash] spec_object A hash of properties that will finally result
+    #                           in the `spec` object exposed to ERB templates
+    # @return [RenderedJobInstance] An object containing the rendering results
+    #                               (when successful)
+    def render(spec_object)
       errors = []
 
-      rendered_templates = @templates.map do |template|
-        job_template_renderer = job_template_renderers[template.name]
+      rendered_templates = @instance_jobs.map do |instance_job|
+        job_template_renderer = job_template_renderers[instance_job.name]
 
         begin
-          job_template_renderer.render(spec)
+          job_template_renderer.render(spec_object)
         rescue Exception => e
           errors.push e
         end
       end
 
       if errors.length > 0
         combined_errors = errors.map{|error| error.message.strip }.join("\n")
-        header = "- Unable to render jobs for instance group '#{spec['job']['name']}'. Errors are:"
+        header = "- Unable to render jobs for instance group '#{spec_object['name']}'. Errors are:"
```

Comment:
Should this be `spec_object['job']['name']` instead of `spec_object['name']`?

Reply:
I’ve deliberately switched to using `spec.name` instead of `spec.job.name`, which holds the same information anyway.

For the reason why, first see: https://bosh.io/docs/jobs/#deprecated-properties-accessors

Indeed, I’ve documented there the `spec.job.*` properties (where “job” is the Bosh v1 name for “intance group” and “ template” the old Bosh v1 name for “job”, so that’s old). To my understanding, these have accidentally leaked in the `spec` object, as there is no reason for `spec.job.*` to be exposed to release authors because the data there are not useful for rendering ERB templates.

That’s why I’ve started favoring `spec.name` over `spec.job.name`. For preparing future removal of `spec.job`.